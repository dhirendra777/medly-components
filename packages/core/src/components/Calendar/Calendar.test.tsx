import { cleanup, fireEvent, render } from '@test-utils';
import React from 'react';
import { Calendar } from './Calendar';
import { CALENDAR_MONTHS } from './constants';
import { getMonthAndDateFromDate } from './helper';

describe('Calendar Component', () => {
    afterAll(cleanup);
    it('should render given date', () => {
        const date = new Date(2020, 0, 1),
            { container } = render(<Calendar id="test-calendar" date={date} onChange={jest.fn()} />);

        expect(container).toHaveTextContent('Jan 2020');
        expect(container).toMatchSnapshot();
    });

    it('should render current month if date is null', () => {
        const { container } = render(<Calendar id="test-calendar" date={null} onChange={jest.fn()} />),
            { month, year } = getMonthAndDateFromDate(new Date());
        expect(container).toHaveTextContent(`${CALENDAR_MONTHS[month]} ${year}`);
    });

    it('should call onChange with expected date', () => {
        const mockOnChange = jest.fn(),
            dateToSelect = new Date(2020, 1, 1),
            { container, getByText, getByTitle } = render(<Calendar id="test-calendar" date={null} onChange={mockOnChange} />);

        fireEvent.click(container.querySelector('#test-calendar-month-selector-input'));
        fireEvent.click(getByText('Feb'));
        fireEvent.click(container.querySelector('#test-calendar-year-selector-input'));
        fireEvent.click(getByText('2020'));
        fireEvent.click(getByTitle(dateToSelect.toDateString()));
        expect(mockOnChange).toHaveBeenCalledWith(dateToSelect);
    });

    it('should render previous month on clicking left arrow when current month is other than Jan', () => {
        const date = new Date(2020, 1, 1),
            { container, getByText } = render(<Calendar id="test-calendar" date={date} onChange={jest.fn()} />);

        fireEvent.click(getByText('<'));
        expect(container).toHaveTextContent(`Jan 2020`);
    });

    it('should render dec month on clicking left arrow when current month is Jan', () => {
        const date = new Date(2020, 0, 1),
            { container, getByText } = render(<Calendar id="test-calendar" date={date} onChange={jest.fn()} />);

        fireEvent.click(getByText('<'));
        expect(container).toHaveTextContent(`Dec 2019`);
    });

    it('should render next month on clicking right arrow when current month is other than Dec', () => {
        const date = new Date(2020, 1, 1),
            { container, getByText } = render(<Calendar id="test-calendar" date={date} onChange={jest.fn()} />);

        fireEvent.click(getByText('>'));
        expect(container).toHaveTextContent(`Mar 2020`);
    });

    it('should render Jan month on clicking right arrow when current month is Dec', () => {
        const date = new Date(2020, 11, 1),
            { container, getByText } = render(<Calendar id="test-calendar" date={date} onChange={jest.fn()} />);

        fireEvent.click(getByText('>'));
        expect(container).toHaveTextContent(`Jan 2021`);
    });

    it('should render 29 days in feb month in leap year', () => {
        const date = new Date(2020, 1, 1),
            { container, getByTitle } = render(<Calendar id="test-calendar" date={date} onChange={jest.fn()} />);

        expect(container).toContainElement(getByTitle('Sat Feb 29 2020'));
    });

    it('should render 28 days in feb month in non leap year', () => {
        const date = new Date(2021, 1, 1),
            { queryByText } = render(<Calendar id="test-calendar" date={date} onChange={jest.fn()} />);

        expect(queryByText('29')).toBeNull();
    });
});
