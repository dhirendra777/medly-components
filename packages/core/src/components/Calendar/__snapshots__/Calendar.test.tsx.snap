// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar Component should render given date 1`] = `
.c4 {
  font-size: 14px;
  font-weight: 500;
  color: inherit;
}

.c5 {
  font-size: 20px;
  font-weight: 400;
  color: inherit;
}

.c20 {
  font-size: 14px;
  font-weight: 400;
  color: inherit;
}

.c22 {
  font-size: 12px;
  font-weight: 400;
  color: inherit;
}

.c17 {
  font-size: 16px;
  cursor: pointer;
}

.c17 * {
  fill: #616161;
}

.c2 {
  background-color: transparent;
  border: none;
  padding: 8px 10px;
  border-radius: 3px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #012040;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2:hover {
  cursor: pointer;
}

.c2:focus {
  outline: none;
}

.c2 .c16 * {
  fill: #012040;
}

.c2:disabled {
  cursor: not-allowed;
}

.c2 .c16 + .c3 {
  margin-left: 10px;
}

.c2 .c3 + .c16 {
  margin-left: 10px;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  margin: 5px;
  overflow: auto;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  box-sizing: border-box;
  background-color: #ffffff;
  padding: 10px;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 4px;
  border: 1px solid #EEEEEE;
  box-shadow: 0 2px 4px 0 rgba(0,0,0,0.2);
}

.c8 {
  margin: 8px 8px 8px 0px;
  display: inline-grid;
  grid-template-columns: minmax(max-content,25%) minmax(auto,80%);
  grid-template-areas: 'label field' 'random description';
  grid-template-columns: auto;
  grid-template-areas: 'field' 'description';
}

.c8 .c25 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c12 {
  margin: 8px 8px 8px 0px;
  display: inline-grid;
  grid-template-areas: 'label' 'field' 'description';
  grid-template-columns: auto;
  grid-template-areas: 'field' 'description';
}

.c12 .c25 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-clip: padding-box;
  box-sizing: border-box;
  border: 1px solid #E0E0E0;
  border-radius: 5px;
  overflow: hidden;
}

.c13:focus-within {
  border-color: #012040;
}

.c13 > * {
  padding: 8px;
  font-size: 14px;
  font-weight: 400;
  color: #616161;
}

.c15 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  outline: 0;
  border: 0;
  cursor: inherit;
  background-color: #ffffff;
}

.c15::-webkit-input-placeholder {
  color: #9E9E9E;
}

.c15::-moz-placeholder {
  color: #9E9E9E;
}

.c15:-ms-input-placeholder {
  color: #9E9E9E;
}

.c15::placeholder {
  color: #9E9E9E;
}

.c15::-webkit-file-upload-button {
  visibility: hidden;
}

.c9 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
  -webkit-flex-direction: column-reverse;
  -ms-flex-direction: column-reverse;
  flex-direction: column-reverse;
}

.c9 .c26 {
  top: calc(100% + 5px);
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

.c11 {
  border: 1px solid #E0E0E0;
  background-color: #ffffff;
  cursor: pointer;
  border-radius: 4px;
  box-sizing: border-box;
  overflow: hidden;
  height: 35px;
  min-width: 150px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c11 .c7 {
  margin: 0px;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c11 .c7 .c14 {
  margin: 0px;
  border: none;
  background-color: transparent;
}

.c11 * {
  cursor: pointer;
}

.c18 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  max-width: 20px;
  margin-right: 8px;
  font-size: 10px;
}

.c1 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c6 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c6 .c10 {
  min-width: unset;
  height: unset;
}

.c6 .c10 input {
  max-width: 50px;
  padding: 5px;
}

.c19 {
  display: grid;
  grid-template: repeat(7,auto) / repeat(7,auto);
}

.c19 > * {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  padding: 8px 10px;
}

.c21 {
  background-color: transparent;
  border: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  border-radius: 5px;
  color: #BDBDBD;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c21:focus {
  outline: none;
}

.c23 {
  background-color: transparent;
  border: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  border-radius: 5px;
  background-color: #FE8D6B;
  color: #ffffff;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c23:focus {
  outline: none;
}

.c24 {
  background-color: transparent;
  border: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  border-radius: 5px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c24:focus {
  outline: none;
}

<div>
  <div
    class="c0"
    id="test-calendar"
  >
    <div
      class="c1"
    >
      <button
        class="c2"
        color="primary"
        type="button"
      >
        <span
          class="c3 c4"
        >
          &lt;
        </span>
      </button>
      <span
        class="c3 c5"
      >
        Jan 2020
      </span>
      <button
        class="c2"
        color="primary"
        type="button"
      >
        <span
          class="c3 c4"
        >
          &gt;
        </span>
      </button>
    </div>
    <div
      class="c6"
    >
      <div
        class="c7 c8"
        id="test-calendar-month-selector-field"
      >
        
        <div
          class="c9"
          id="medly-popover-wrapper"
        >
          <div
            class="c10 c11"
          >
            <div
              class="c7 c12"
              id="test-calendar-month-selector-input-field"
            >
              
              <div
                class="c13"
              >
                <input
                  autocomplete="off"
                  class="c14 c15"
                  id="test-calendar-month-selector-input"
                  placeholder="Please Select . . ."
                  type="text"
                  value="Jan"
                />
              </div>
              
            </div>
            <svg
              class="c16 c17 c18"
              height="1em"
              theme="[object Object]"
              viewBox="0 0 12 8"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M9.88 1.29L6 5.17 2.12 1.29A.996.996 0 1 0 .71 2.7L5.3 7.29c.39.39 1.02.39 1.41 0L11.3 2.7a.996.996 0 0 0 0-1.41c-.39-.38-1.03-.39-1.42 0z"
              />
            </svg>
          </div>
        </div>
        
      </div>
      <div
        class="c7 c8"
        id="test-calendar-year-selector-field"
      >
        
        <div
          class="c9"
          id="medly-popover-wrapper"
        >
          <div
            class="c10 c11"
          >
            <div
              class="c7 c12"
              id="test-calendar-year-selector-input-field"
            >
              
              <div
                class="c13"
              >
                <input
                  autocomplete="off"
                  class="c14 c15"
                  id="test-calendar-year-selector-input"
                  placeholder="Please Select . . ."
                  type="text"
                  value="2020"
                />
              </div>
              
            </div>
            <svg
              class="c16 c17 c18"
              height="1em"
              theme="[object Object]"
              viewBox="0 0 12 8"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M9.88 1.29L6 5.17 2.12 1.29A.996.996 0 1 0 .71 2.7L5.3 7.29c.39.39 1.02.39 1.41 0L11.3 2.7a.996.996 0 0 0 0-1.41c-.39-.38-1.03-.39-1.42 0z"
              />
            </svg>
          </div>
        </div>
        
      </div>
    </div>
    <div
      class="c19"
    >
      <span
        class="c3 c20"
      >
        Su
      </span>
      <span
        class="c3 c20"
      >
        Mo
      </span>
      <span
        class="c3 c20"
      >
        Tu
      </span>
      <span
        class="c3 c20"
      >
        We
      </span>
      <span
        class="c3 c20"
      >
        Th
      </span>
      <span
        class="c3 c20"
      >
        Fr
      </span>
      <span
        class="c3 c20"
      >
        Sa
      </span>
      <button
        class="c21"
        title="Sun Dec 29 2019"
        type="button"
      >
        <span
          class="c3 c22"
        >
          29
        </span>
      </button>
      <button
        class="c21"
        title="Mon Dec 30 2019"
        type="button"
      >
        <span
          class="c3 c22"
        >
          30
        </span>
      </button>
      <button
        class="c21"
        title="Tue Dec 31 2019"
        type="button"
      >
        <span
          class="c3 c22"
        >
          31
        </span>
      </button>
      <button
        class="c23"
        title="Wed Jan 01 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          1
        </span>
      </button>
      <button
        class="c24"
        title="Thu Jan 02 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          2
        </span>
      </button>
      <button
        class="c24"
        title="Fri Jan 03 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          3
        </span>
      </button>
      <button
        class="c24"
        title="Sat Jan 04 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          4
        </span>
      </button>
      <button
        class="c24"
        title="Sun Jan 05 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          5
        </span>
      </button>
      <button
        class="c24"
        title="Mon Jan 06 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          6
        </span>
      </button>
      <button
        class="c24"
        title="Tue Jan 07 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          7
        </span>
      </button>
      <button
        class="c24"
        title="Wed Jan 08 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          8
        </span>
      </button>
      <button
        class="c24"
        title="Thu Jan 09 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          9
        </span>
      </button>
      <button
        class="c24"
        title="Fri Jan 10 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          10
        </span>
      </button>
      <button
        class="c24"
        title="Sat Jan 11 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          11
        </span>
      </button>
      <button
        class="c24"
        title="Sun Jan 12 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          12
        </span>
      </button>
      <button
        class="c24"
        title="Mon Jan 13 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          13
        </span>
      </button>
      <button
        class="c24"
        title="Tue Jan 14 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          14
        </span>
      </button>
      <button
        class="c24"
        title="Wed Jan 15 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          15
        </span>
      </button>
      <button
        class="c24"
        title="Thu Jan 16 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          16
        </span>
      </button>
      <button
        class="c24"
        title="Fri Jan 17 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          17
        </span>
      </button>
      <button
        class="c24"
        title="Sat Jan 18 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          18
        </span>
      </button>
      <button
        class="c24"
        title="Sun Jan 19 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          19
        </span>
      </button>
      <button
        class="c24"
        title="Mon Jan 20 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          20
        </span>
      </button>
      <button
        class="c24"
        title="Tue Jan 21 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          21
        </span>
      </button>
      <button
        class="c24"
        title="Wed Jan 22 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          22
        </span>
      </button>
      <button
        class="c24"
        title="Thu Jan 23 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          23
        </span>
      </button>
      <button
        class="c24"
        title="Fri Jan 24 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          24
        </span>
      </button>
      <button
        class="c24"
        title="Sat Jan 25 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          25
        </span>
      </button>
      <button
        class="c24"
        title="Sun Jan 26 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          26
        </span>
      </button>
      <button
        class="c24"
        title="Mon Jan 27 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          27
        </span>
      </button>
      <button
        class="c24"
        title="Tue Jan 28 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          28
        </span>
      </button>
      <button
        class="c24"
        title="Wed Jan 29 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          29
        </span>
      </button>
      <button
        class="c24"
        title="Thu Jan 30 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          30
        </span>
      </button>
      <button
        class="c24"
        title="Fri Jan 31 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          31
        </span>
      </button>
      <button
        class="c21"
        title="Sat Feb 01 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          1
        </span>
      </button>
      <button
        class="c21"
        title="Sun Feb 02 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          2
        </span>
      </button>
      <button
        class="c21"
        title="Mon Feb 03 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          3
        </span>
      </button>
      <button
        class="c21"
        title="Tue Feb 04 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          4
        </span>
      </button>
      <button
        class="c21"
        title="Wed Feb 05 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          5
        </span>
      </button>
      <button
        class="c21"
        title="Thu Feb 06 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          6
        </span>
      </button>
      <button
        class="c21"
        title="Fri Feb 07 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          7
        </span>
      </button>
      <button
        class="c21"
        title="Sat Feb 08 2020"
        type="button"
      >
        <span
          class="c3 c22"
        >
          8
        </span>
      </button>
    </div>
  </div>
</div>
`;
